soc/usbexecd.c

```diff
@@ -0,0 +1,598 @@
+int pipebuf_common_init(pipebuf_t* pb, int id, int role, uint32_t events, unsigned bufend) {
+    // Assign all fields to the pipe buffer
+    pb->id = id;
+    pb->role = role;
+    pb->eof = false;
+    pb->startpos = pb->endpos = pb->bufcount = 0;
+    pb->credit = 0;
+
+    // Allocate memory for file descriptors
+    int pipefd[2];
+    // Create the pipes
+    int r = pipe(pipefd);
+    // Verify it worked
+    if (r < 0) {
+        fatal("pipe failed: %s", strerror(errno));
+    }
+
+    // Set the proper read or write pipe file descriptor
+    pb->fd = pipefd[bufend];
+
+    // Create an epoll_event related to that file descriptor
+    struct epoll_event evt;
+    evt.data.ptr = pb;
+    evt.events = events;
+
+    // Register the event
+    r = epoll_ctl(epfd, EPOLL_CTL_ADD, pb->fd, &evt);
+    // Verify it worked
+    if (r < 0) {
+        fatal("epoll_ctl failed: %s", strerror(errno));
+    }
+
+    // Return the other file descriptor?
+    return pipefd[!bufend];
+}
+
+// Closing functionality for read and write pipe buffers
+void pipebuf_common_close(pipebuf_t* pb) {
+    // If the file descriptor exists (which it should)
+    if (pb->fd) {
+        // Close it
+        close(pb->fd);
+        // Set it back to 0
+        pb->fd = 0;
+    }
+    // Reset other fields
+    pb->startpos = pb->endpos = pb->bufcount = 0;
+    pb->eof = true;
+
+    // Tell the other end of the pipe that this fd has been closed
+    send_header(CMD_CLOSE_CONTROL + pb->role, pb->id, 0, 0);
+}
+
+int pipebuf_in_init(pipebuf_t* pb, int id, int role) {
+    return pipebuf_common_init(pb, id, role, EPOLLIN, 0);
+}
+
+// We have data coming in to stdout or stderr
+// and need to write it to the CLI
+void pipebuf_in_handle(pipebuf_t* pb) {
+    int len = 255;
+    // if (pb->credit < len) len = pb->credit;
+    char buf[255];
+    len = read(pb->fd, buf, len);
+    if (len < 0) {
+        if (errno == EAGAIN) {
+            return;
+        }
+        fatal("error in pipe read: %s", strerror(errno));
+    }
+
+    send_header(CMD_WRITE_CONTROL + pb->role, pb->id, 0, len);
+    write(sock, buf, len);
+
+    // pb->credit -= len;
+    // if (pb->credit == 0) {
+        // disable poll
+        // FOr example, the child process has stuff to write and the 
+        // CLI isn't able to read. Stop notifying so it doesn't loop forever
+    // }
+}
+
+// 
+void pipebuf_in_ack(pipebuf_t* pb, size_t acksize) {
+    if (pb->credit == 0 && acksize > 0) {
+        // enable poll
+    }
+    pb->credit += acksize;
+}
+
+void pipebuf_in_close(pipebuf_t* pb) {
+    pipebuf_common_close(pb);
+}
+
+
+int pipebuf_out_init(pipebuf_t* pb, int id, int role) {
+    int fd =  pipebuf_common_init(pb, id, role, EPOLLIN, 1);
+
+    send_header(CMD_ACK_CONTROL + pb->role, pb->id, 255, 0);
+    pb->credit = 255;
+
+    return fd;
+}
+
+// Stdin or StdCtl has bytes that need to be read
+void pipebuf_out_handle(pipebuf_t* pb) {
+    uint8_t max_bytes = 255;
```

Comment:
I don't think this function is quite right. The stdin and control file descriptors will never be readable because these are the write sides. Instead, it should be waiting for EPOLLOUT and this callback for that event should give credit to the CLI to send more data.